RDART-969: Keypath filtering on collections

papafe opened the pull request:
This PR adds support for keypath filtering on collections

TODO:
- [x] Add support for lists
- [x] Add support for sets
- [x] Add support for dictionaries
- [x] Add basic tests for lists, sets, dictionaries (bulk of the tests will be on results). This includes verifying we can't use keypath filtering on collections that do not contain realm objects
- [x] Changelog

papafe commented:
@nielsenko It needs a couple of things (like the changelog), but I think it can already be reviewed. 
I've put the bulk of the tests for the keypath filtering on results, and I've left some basic tests for the other collections (lists, sets, dictionaries), so we don't repeat the same tests.
I have left some todos around about things I am not sure / we need to discuss, but functionally (almost) everything should work

papafe commented on packages/realm_dart/test/realm_object_test.dart:
```diff
@@ -522,7 +522,7 @@ void main() {
     });
 
-    test('empty list gives default subscriptions', () async {
+    test('empty list gives no subscriptions', () async {
```
Not sure if in this case we prefer to throw an exception or not

nielsenko reviewed (changes requested):
Look mostly good to me. Most of my comments are more suggestions, but the the one about `import 'dart:ffi';` is a blocker 😄

papafe commented on CHANGELOG.md:
```diff
@@ -51,8 +51,21 @@
 * Allow the realm package, and code generated by realm_generator to be included when building 
-  for web without breaking compilation. (Issue [#1374](https://github.com/realm/realm-dart/issues/1374),
-  PR [#1713](https://github.com/realm/realm-dart/pull/1713)). This does **not** imply that realm works on web!
+  for web without breaking compilation. (Issue [#1374](https://github.com/realm/realm-dart/issues/1374))
```
Not really, sorry, I messed up the conflict it seems 🤦

papafe commented on packages/realm_dart/lib/src/handles/native/object_handle.dart:
```diff
@@ -138,7 +138,7 @@ class ObjectHandle extends RootedHandleBase<realm_object> implements intf.Object
   NotificationTokenHandle subscribeForNotifications(NotificationsController controller, [List<String>? keyPaths]) {
     return using((arena) {
-      final kpNative = buildAndVerifyKeyPath(keyPaths);
+      final kpNative = root.buildAndVerifyKeyPath(keyPaths, classKey);
```
Yes we can, I'll do it

papafe commented on packages/realm_dart/lib/src/results.dart:
```diff
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import 'dart:async';
+import 'dart:ffi';
```
Agreed

papafe commented on packages/realm_dart/test/realm_object_test.dart:
```diff
@@ -558,7 +559,7 @@ void main() {
 
       subscription.cancel();
-    });
+    }, skip: true);
```
There is a comment on top of the method

nielsenko replied:
.. missed that

papafe commented on packages/realm_dart/lib/src/map.dart:
```diff
@@ -273,12 +287,23 @@ extension RealmMapInternal<T extends Object?> on RealmMap<T> {
+    if (keyPaths != null) {
+      this.keyPaths = keyPaths;
+
+      if (keyPaths.any((element) => element.isEmpty || element.trim().isEmpty)) {
+        throw RealmException("A key path cannot be empty or consisting only of white spaces");
+      }
```
We can definitely move it to `verifyKeyPath`, and yes I can improve the wording. 
But no, core won't complain, and I thought it would be better to throw an exception for this.

papafe commented on packages/realm_dart/lib/src/list.dart:
```diff
@@ -42,8 +42,12 @@ abstract class RealmList<T extends Object?> with RealmEntity implements List<T>
   Stream<RealmListChanges<T>> get changes;
+
+  /// Allows listening for changes when the contents of this collection changes on one of the provided [keyPaths].
+  Stream<RealmListChanges<T>> changesFor([List<String>? keyPaths]);
```
That would actually make sense. 
The only "issue" I see is that now `changes` is defined as `changesFor(null)` to avoid repetitions. 
If we move `changesFor` to an extension method then we'd need to have some duplication (admittedly not a biggie though). It also seems Dart does not have protected methods, so I don't know if I can go around this with some other language constructs. Any suggestion?

nielsenko replied:
You could make a private `_changesFor` that both the `changesFor` extension method, and `changes` member delegates to?

papafe replied:
Yeah, I was thinking something like that. In my mind it sounded uglier than it would probably be, I'll give it a try 😁

papafe replied:
I've done it in the last commit

papafe commented on packages/realm_dart/lib/src/map.dart:
```diff
@@ -33,6 +33,9 @@ abstract class RealmMap<T extends Object?> with RealmEntity implements MapBase<S
   /// Allows listening for changes when the contents of this collection changes.
   Stream<RealmMapChanges<T>> get changes;
+
+  /// Allows listening for changes when the contents of this collection changes on one of the provided [keyPaths].
```
Added more explanation on the extension method

papafe commented:
@nielsenko I think I've addressed all the comments. 
A couple of other things I've noticed, and that we don't necessarily need to address here, but I thought it would be worth writing down.

We have some inconsistencies with naming:
1. In the errors for unmanaged sets we call them `RealmSets`, while we call lists just `Lists` (same for maps)
2. Regarding notification controllers, we have `ListNotificationController` but `RealmSetNotificationController`.
3. Regarding tests, we have `list_test` but we have `realm_set_test`/`realm_map_test`
4. Maybe we should have a different suffix between the native handle implementation and the interface files. just having a look is a little bit confusing sometimes.

Some other considerations (that probably require some work):
1. Maybe we can move the keypath property to the base `NotificationController`,  instead of having it defined in the various specific classes
2. Maybe we can have a base class for all collections (results, lists, sets, maps). This way we could move `changes` and some other common things there. The collection handle themselves could have a common base class.

nielsenko commented:
In my opinion..

> 1. In the errors for unmanaged sets we call them `RealmSets`, while we call lists just `Lists` (same for maps)

Going forward we should try to reduce the use of _realm_ prefix to where it is needed. This means only on public facing classes, that are likely to collide with other class names (`RealmSet`, `RealmList`, `RealmMap`, etc.). Internally we should avoid unless it makes code more readable.

> 2. Regarding notification controllers, we have `ListNotificationController` but `RealmSetNotificationController`.

It should just be `SetNotificationController`.

> 3. Regarding tests, we have `list_test` but we have `realm_set_test`/`realm_map_test`

it should just be `set_test.dart`.

> 4. Maybe we should have a different suffix between the native handle implementation and the interface files. just having a look is a little bit confusing sometimes.

I'm a bit reluctant on this. For each handle, there is the interface, and two implementations.. I would prefer not to suffix the implementations with `Native` and `Web` (or whatever)

> Some other considerations (that probably require some work):
> 
> 1. Maybe we can move the keypath property to the base `NotificationController`,  instead of having it defined in the various specific classes

> 2. Maybe we can have a base class for all collections (results, lists, sets, maps). This way we could move `changes` and some other common things there. The collection handle themselves could have a common base class.

The current C-api don't really give us much benefit, but yes.

However, let us land this PR as is, and get started on the weekend 😄 .. we can always address this later.

nielsenko reviewed (approved):
LGTM! Great work 👍

papafe commented:
I'll merge this as soon as the CI agrees
